Site Assembler - Write Flow (dotcom patterns only)

miksansegundo opened the pull request:
#### Proposed Changes

* Enables the feature flag "**pattern-assembler/write-flow**" in calypso and development environments to:
  * Show the assembler CTA in the Write flow.
  * Add query patterns from `dotcompatterns` in the sections list. 
* Render the query pattern previews using posts from `blankcanvas3demo` site.
* Pass the params `trim_content: false` and `post_source_site_id` to `/theme-setup` to copy all the posts (4) from `blankcanvas3demo` site into the user's site at the end of the flow. 

Note that changes made in the posts on `blankcanvas3demo` site will affect the following:
- the previews of query patterns in the assembler  
- the posts created on the user's site at the end of the assembler flow

https://user-images.githubusercontent.com/1881481/215760944-ea9a4567-bd7d-4353-b887-af67b760621c.mov

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Test using a non-a8c account (we do this to check that all users can request all the endpoints without having permission issues). 
- Create a new site from the admin or visit `/start`.
- Select a domain and a plan.
- In the goals screen:
  - in other environments different than development, or calypso, you'll need to add this param in the URL `&flags=pattern-assembler/write-flow` and press enter.  
  - then select Write & Publish.
- Continue until the Design picker and scroll down to click `Start designing` on the CTA. If you don't see it, check that the feature flag is in the URL.
- Click `Add sections` to choose query patterns. If you hover them, you'll see their category name `Posts` in a tooltip.
- Check that the previews use 4 posts with image placeholders as featured images.
- Click `Continue` to complete the flow and you'll land on the editor.
- Check that the previews in the editor and the site front-end match the previews in the assembler.
- Check that four posts with feature images have been created on your site.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [X] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [X] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [X] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [X] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [X] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/72565

miksansegundo committed:
Added query pattern ids in the posts category

miksansegundo committed:
Show the assembler CTA in build and write flows

miksansegundo committed:
Show the assembler CTA in build and write flows

miksansegundo committed:
Use posts from dotcomverticalposts

miksansegundo committed:
Add support to disable the headstart setting that trims the content and limits the posts

bangank36 commented:
Some early feedbacks for the PR
**Query pattern on preview**
<details>
  <summary>Expand</summary>
  
| Pattern ID | Preview | Notes |
|-------------------------|----------------------|----------------------------|
|         5645  <br> Four Recent Blog Posts              |           ![LjJFog.png](https://user-images.githubusercontent.com/10071857/215376667-d0655831-951d-4238-a076-ed4a95712a64.png)           |                            |
|         1784 <br> Recent Posts                |           ![jMatMx.png](https://user-images.githubusercontent.com/10071857/215376953-27e468f4-9af8-4732-a6fe-35286ce00ee3.png)         |                            |
|         8421 <br> Grid of posts 2x3                |          ![7ug2AA.png](https://user-images.githubusercontent.com/10071857/215377680-9c5b8913-f67b-47b4-99bc-199c5040d8a0.png)        |                            |
|         8435 <br> Grid of Posts 3x2                |          ![ca2cYc.png](https://user-images.githubusercontent.com/10071857/215378606-33f891d5-5de8-4edc-b01a-c697e98201ee.png)      |                            |
|         7996 <br> Grid of Posts 4x2               |          ![zjNESP.png](https://user-images.githubusercontent.com/10071857/215378029-54be9c2c-16a3-4571-8c42-d27cee05b052.png)         |                            |
|         8437 <br> List of posts              |           ![VL2tcV.png](https://user-images.githubusercontent.com/10071857/215378115-41361a15-e23d-411d-a11f-9ef16dde4315.png)        |                            |
|         3213 <br> Latest podcast episodes            |          ![tOZhUU.png](https://user-images.githubusercontent.com/10071857/215377942-d81c14a1-beaf-4145-bd2b-c2d6e75b6603.png)     |                            |
</details>

**Number of Posts on new site**
![Z4UJn9.png](https://user-images.githubusercontent.com/10071857/215378454-4fbbc0e0-65f8-4f1c-8160-5dd479affed8.png)

**On the site editor**
It's not look so good though 
![ih7Pd5.png](https://user-images.githubusercontent.com/10071857/215378524-86968e91-0839-4cd6-80da-59379afa47b3.png)

miksansegundo commented:
Thanks so much for adding screenshots @bangank36. 

I noticed and identified the cause of those issues except for the issue with the featured image placeholders in the editor. I'll try to find an open issue in the Gutenberg repo.

miksansegundo committed:
Disable trim_content to create posts in the user site without limit

miksansegundo commented:
@bangank36 can you sandbox D99603-code and test again, please?

I created 4 posts using the placeholder image as the featured image on `blankcanvas3demo`. Let's start with these and then see if we want more.

<img width="924" alt="Screenshot 2566-01-30 at 14 17 09" src="https://user-images.githubusercontent.com/1881481/215411836-ddb6cf72-11d9-40c7-8a73-821884347d55.png">

Also, I added the following changes to this PR:

- [Disable trim_content to create posts in the user site without limit](https://github.com/Automattic/wp-calypso/pull/72748/commits/54186ba88d0824fb521006a8eeb2de4c81e6f0f4)
  - This change removes the limit of posts created on the user's site. All the posts (4) on the site `blankcanvas3demo` will be copied to the user's site.
- [Use blank canvas 3 demo site for rendering pattern previews](https://github.com/Automattic/wp-calypso/pull/72748/commits/17cc70ee11f40928ace85f4c522e8e65bc93d8aa)
  - This change makes the query pattern previews use all the posts (4) from `blankcanvas3demo`.

Pending TODO:

- Pass a new param to `/theme-setup` to generate the headstart annotation with posts from `blankcanvas3demo` rather than from `dotcomverticalposts`.

cc: @autumnfjeld @arthur791004

bangank36 commented:
> can you sandbox D99603-code and test again, please?

On it!

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx:
```diff
@@ -100,9 +100,9 @@ const UnifiedDesignPickerStep: Step = ( { navigation, flow } ) => {
+			// Extract the blank canvas design first and then insert it into the last one for the build and write intents
 			const blankCanvasDesign = allDesigns.static.designs.splice( blankCanvasDesignOffset, 1 );
-			if ( isEnabled( 'signup/design-picker-pattern-assembler' ) && intent === SiteIntent.Build ) {
+			if ( isEnabled( 'signup/design-picker-pattern-assembler' ) && intent !== SiteIntent.Sell ) {
```
I think the following one would be safer

```js
const SITE_ASSEMBLER_AVAILABLE_INTENTS = [
  SiteIntent.Build,
  SiteIntent.Write,
];

...

SITE_ASSEMBLER_AVAILABLE_INTENTS.includes( intent )
```

arthur791004 commented on packages/data-stores/src/site/actions.ts:
```diff
@@ -398,7 +398,7 @@ export function createActions( clientCreds: WpcomClientCredentials ) {
 
 		const themeSetupOptions: ThemeSetupOptions = {
-			trim_content: true,
+			trim_content: options?.trimContent !== undefined ? options?.trimContent : true,
```
[nit] we can use nullish operator here: `options?.trimContent ?? true`

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-assembler-container.tsx:
```diff
@@ -18,11 +19,11 @@ const PatternAssemblerContainer = ( {
-			// For example, site logo, site title, site tagline, posts.
-			siteId={ siteId }
+			// Use this site to render the posts of query patterns.
+			siteId={ POST_SITE_ID }
```
Calling it `PLACEHOLDER_SITE` might be better as it would also affect the site logo (and even the site title and tagline if the `siteInfo` is not provided). What do you think?

miksansegundo committed:
Code improvements

miksansegundo commented:
> This PR modifies the release build for editing-toolkit

Is that because of the change in `packages/data-stores/src/site/actions.ts`?

arthur791004 commented:
> Is that because of the change in packages/data-stores/src/site/actions.ts?

yes

miksansegundo committed:
Pass is_assembler_flow to /theme-setup to copy posts from blank-canvas-3 demo site

miksansegundo commented:
I updated this PR to pass the param `is_assembler_flow: true` to `/theme-setup` to copy the posts from `blankcanvas3demo` site, rather than from `dotcomverticalposts`. For testing, now it is required to sandbox D99790-code too.

I have updated the PR description. There aren't pending tasks!

miksansegundo committed:
Add is_assembler_flow to /theme-setup tests

arthur791004 commented:
This PR looks good, and let's wait for D99790-code 🙂

miksansegundo committed:
Pass posts_source_site_id in /theme-setup to copy starter content from blank-canvas-3 demo site

miksansegundo commented:
This PR can be tested without sandboxing the API because we merged the following diffs: 
- D99603-code to make sure the posts created on the user's site have featured images
- D99790-code to copy the posts from `blankcanvas3demo` on the user's site at the end of the flow

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/patterns-data.ts:
```diff
@@ -253,6 +254,41 @@ const useSectionPatterns = () => {
+			{
+				id: 5645,
+				name: 'Four Recent Blog Posts',
+				category: posts,
```
Question: Do we want to show these patterns in the build flow?

miksansegundo replied:
I think they don't hurt and wouldn't be a blocker anyways. Nobody has mentioned that they shouldn't be shown in other flows. I believe we want them in all flows because they are a core piece of WP.

miksansegundo replied:
@autumnfjeld, we could consider filtering patterns, or even categories, based on the user's goals/intent in the next iterations.

autumnfjeld replied:
> I think they don't hurt and wouldn't be a blocker anyways. Nobody has mentioned that they shouldn't be shown in other flows. I believe we want them in all flows because they are a core piece of WP.

I agree.

miksansegundo committed:
rebase

arthur791004 commented on packages/data-stores/src/site/actions.ts:
```diff
@@ -398,7 +398,8 @@ export function createActions( clientCreds: WpcomClientCredentials ) {
 		const themeSetupOptions: ThemeSetupOptions = {
-			trim_content: true,
+			trim_content: options?.trimContent ?? true,
+			posts_source_site_id: options?.posts_source_site_id,
```
[nit] We can set the `posts_source_site_id` property when it's provided, and then we don't need to update the test case.

arthur791004 reviewed (approved):
Tested and looks good to me!

miksansegundo committed:
Code improvement

autumnfjeld commented:
The posts block I chose in the PA:

<img width="1268" alt="image" src="https://user-images.githubusercontent.com/5665959/216244017-e562edd7-1f82-460e-9640-3fc6a9022cc5.png">

How it looks in the Site Editor
<img width="1310" alt="image" src="https://user-images.githubusercontent.com/5665959/216244748-2d6b1365-55de-46cc-b55a-a6d3400d3acd.png">



How it looks on the front end
🚫  I see this [modal](https://github.com/Automattic/wp-calypso/issues/72093): it conflicts with the Tour and I think is also not supposed to show on page load, though I don't know for sure.  
![image](https://user-images.githubusercontent.com/5665959/216247572-333f681f-ab82-4194-b1a9-5737513f8144.png)


I see the modal again when I click save and when I click Next Steps it takes me to the launch pad.


https://user-images.githubusercontent.com/5665959/216247795-bd411ea2-ffdf-4a48-88a7-32f3e478fbe0.mp4



So we'll need to fix this before publishing the CfT.

miksansegundo committed:
Add feature flag pattern-assembler/write-flow

miksansegundo committed:
Enable feature flag pattern-assembler/write-flow on horizon and development envs

miksansegundo commented:
I added the feature flag `pattern-assembler/write-flow` and enabled it on `development` and `horizon` envs.

miksansegundo commented:
> I see this https://github.com/Automattic/wp-calypso/issues/72093: it conflicts with the Tour and I think is also not supposed to show on page load, though I don't know for sure.

We haven't reproduced that issue. We believe that was a false positive and confirmed that the modal shows up only when you publish a post or save changes on the site editor, and the LaunchPad checklist is completed. 

See the [implementation](https://github.com/Automattic/jetpack/blob/444a237e732a0405ac3ffb0ec0585469e465d06a/projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js#L70), and more testing details on pbxlJb-3rg-p2.
